scipy/optimize/optimize.py

```diff
@@ -51,26 +51,30 @@
 
 
 class MemoizeJac(object):
-    """ Decorator that caches the value gradient of function each time it
-    is called. """
+    """ Decorator that caches the return values of a function returning `(fun, grad)`
+        each time it is called. """
 
     def __init__(self, fun):
         self.fun = fun
         self.jac = None
+        self._value = None
         self.x = None
 
+    def _compute_if_needed(self, x, *args):
+        if not np.all(x == self.x) or self._value is None or self.jac is None:
+            self.x = np.asarray(x).copy()
+            fg = self.fun(x, *args)
```

Comment:
Does 
```
self._value, self.jac = self.fun(x, *args)
```
look cleaner?

Reply:
absolutely. Technically this would also be a break of backwarts compatibility, since the old code allowed the function to return `fun, jac, *random_stuff`, but I guess that is not a problem?

Reply:
Ahh, good point. Leave as is then.